pySDC/projects/Monodomain/sweeper_classes/exponential_runge_kutta/imexexp_1st_order.py

```diff
@@ -0,0 +1,368 @@
+        self.QI = self.get_Qdelta_implicit(coll=self.coll, qd_type=self.params.QI)
+        self.delta = np.diagonal(self.QI)[1:]
+
+        self.lmbda = None
+
+        self.lambda_and_phi_outdated = True
+
+        # Compute weights w such that PiQ^(k)(0) = sum_{j=0}^{M-1} w[k,j]*Q[j], k=0,...,M-1
+        # Used to express the derivatives of a polynomial in x=0 in terms of the values of the polynomial at the collocation nodes
+        M = self.coll.num_nodes
+        c = self.coll.nodes
+        self.w = fornberg.fd_weights_all(c, 0.0, M - 1)
+
+        # Define the quadature rule for the evaluation of the phi_i(z) functions. Indeed, we evaluate them as integrals in order to avoid round off errors.
+        phi_num_nodes = 5  # seems to be enough in most cases
+        self.phi_coll = CollBase(num_nodes=phi_num_nodes, tleft=0, tright=1, node_type='LEGENDRE', quad_type='GAUSS')
+
+    def phi_eval_lists(self, P, factors, indeces, phi, lmbda, update_non_exp_indeces=True):
+        """
+        Evaluate the phi_k functions at the points factor_i*lmbda
+
+        Arguments:
+            P: problem class
+            factors: list of factors to multiply lmbda with. len(factors)=len(self.num_nodes)
+            indeces: list of indeces k for the phi_k functions
+            phi: list of lists of dtype_u: some space to store the results
+            lmbda: dtype_u: the value of lmbda
+            update_non_exp_indeces: bool: if True, the phi functions are also evaluated at the non-exponential indeces. Hence, where lambda=0 and thus we return phi_k(0) (using analytical value)
+
+        Returns:
+            list of lists of dtype_u: [[phi_k(factor_i*lambda) for i in range(len(factors))] for k in indeces]
+
+        """
+
+        N_fac = len(factors)
+        N_ind = len(indeces)
+
+        if phi is None:
+            # make some space
+            phi = [[P.dtype_u(init=P.init, val=0.0) for i in range(N_fac)] for j in range(N_ind)]
+        else:
+            # zero out the provided space
+            for n in range(N_fac):
+                for m in range(N_ind):
+                    phi[m][n][P.rhs_exp_indeces] = 0.0
+
+        factorials = scipy.special.factorial(np.array(indeces) - 1)
+        # the quadrature rule is used to evaluate the phi functions as integrals. This is not the same as the one used in the ESDC method!!!!
+        c = self.phi_coll.nodes
+        b = self.phi_coll.weights
+
+        # iterate only over the indeces of the problem having an exponential term
+        # hence we update only the subvectors phi_k(factor*lambda)_i in the vector of vectors phi_k(factor*lambda)
+        for i in P.rhs_exp_indeces:
+            # iterate over all the factors
+            for n in range(N_fac):
+                factor = factors[n]
+                # compute e^((1-c_j)*factor*lmbda) for nodes c_j on the quadrature rule
+                exp_terms = [np.exp(((1.0 - c[j]) * factor) * lmbda[i]) for j in range(self.phi_coll.num_nodes)]
```

Comment:
This line for instance is slow, with lmbda a mesh instance of shape (19,512), for instance. With 19 the number of model unknowns and 512 the mesh size

Reply:
I think we need a list comprehension here, so that Numpy can do it's magic. Do-loops are evil in Python.

Reply:
If I understand correctly, the problem is the outer loop over `P.rhs_exp_indeces`. You are solving different components with different methods. So `dtype_u` should not be `mesh` but some sort of `MultiComponentMesh` as well. I guess, specific to the monodomain equation, you would have
```
class MonodomainMesh(MultiComponentMesh):
    components = ['V', 'z'] 
```
or something like that.
By using masks, you should be much faster than with an explicit for loop, that is to say the code in the sweeper should be 
```
for n in range(N_fac):
    factor = factors[n]
    # compute e^((1-c_j)*factor*lmbda) for nodes c_j on the quadrature rule
    exp_terms = [np.exp(((1.0 - c[j]) * factor) * lmbda.z) for j in range(self.phi_coll.num_nodes)]
    ...
    phi[m][n].z = np.exp(factor * lambda.z)
    ...
```
At this point, you will run into the issue that you want, in general, different shapes for V and z, I assume. But this can surely be dealt with somehow.

Now that I wrote all of this, I am not sure that is what is slow, though. After all, the for loop is run only over few indices. I find this very confusing, to be honest. Can you check if `lambda[i]` owns its memory by printing `lambda[I].flags`? We don't want it to, as we want a view on the original data. Can you send me the profile? Did you use cProfile?

Reply:
But this is O(indices * num_nodes * N_fac). The hope would be to get this into one big numpy expression with lots of `[:]` to get this to full speed (with threading).

Reply:
I did like Robert suggested :)

Reply:
Great job!!